### PR TITLE
FIX: Composite touchscreen controls not firing action after enabling (ISXB-98)

### PR DESCRIPTION
**Pull request by jamesmcgill**
### Description

After the first enabling of an action map that is bound to composite touchscreen controls, the action would not fire if the touchscreen was not in the default state at this point. Which would happen if the screen was ever touched prior to enabling the action. (This is because Touchscreens are never in the default state if they were ever used. E.g. the position never returns to default and is essentially always actuated forever).

The issue was due to sorting behaviour introduced in https://github.com/Unity-Technologies/InputSystem/pull/1405
This sorting would reverse the order of the Touchscreen controls which just happened to work before because the Press Control was first in the list (after sorting everything was reversed; device order was Touch 4, Touch3, Touch 2 etc, and controls within are reversed, with Pressed now last).
This fragile behaviour of the Touchscreen controls always being actuated and the order mattering is not addressed in this PR. What this PR does, is restore the previous ordering but still doing the sorting. As sorting by composite complexity is required for the composite PR #1405 to work.  

### Changes made

- Change sorting to be only based on complexity but otherwise be stable sorting (e.g. no reversing of controls)
- Revert unit test changes that were done in PR #1405 which were only due to the ordering change.

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1540) <br/>All committers have signed the CLA.

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs:10051]: Which demo project? Bug repro project? If this is the case it might be good to include bug issue number in case we need to trace it at some point. I noticed its in the changelog, but maybe add a link to public issue tracker here?

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs:10109]: Excellent you added this because I was writing a comment about it where it said "demo" project, either leave it or call it "repro project" for clarity

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs]: TODO intended?, not sure what this is referencing?

File: Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs]: Not really following what is going on here, what are these rshift by 48 and mask byte? Maybe wort a comment what we have stored in MSB byte or maybe this is documented on the member?

**Review by ekcoh — APPROVED** [review]
@jamesmcgill Great description of the issue and fix in this PR and walking the reader through what was causing the problem. Impressed you were able to decipher this issue. Is there something regarding this behavior we should add or are missing in docs in your opinion?

File: Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs]: I guess its some kind of priority combined with index?

Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs]: I found that monitorIndex isn't really an index but some sort of bit packed field.
It's in InputActionState::ToCombinedMapAndControlAndBindingIndex().

Another alternative here would be to make InputActionState::SplitUpMapAndControlAndBindingIndex  public instead of private. And then extend it to pull out the complexity too. Then the code above would be simpler to read. The downside would be that there would be a bunch of extra unpacking and temporary variables needed here.  Probably this is packed for performance reasons so I wanted the simplest code.

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (jamesmcgill) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: I actually don't know either. @jimon wrote the test. But probably I should tidy it up a bit before merging.

**Comment by jamesmcgill** [comment]
Thanks for the review @ekcoh !
I've made some changes (cleanup the test, put the bit unpacking in a function, CHANGELOG etc.)

I found one place that the sorting was mentioned in the docs and I edited that part. But generally I think the sorting is an internal detail so I don't think we will need to add anything (especially as this PR just partially restores the previous behaviour rather than add something new).

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs]: I don't think these param comments are correct. The arguments are input only, and are not changed by the method. These three values are packed together into the returned long. Ah, I see. The same comments are on the SplitUpMapAndControlBindingIndex method. Copy paste error?

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs]: Modifiers are in the wrong order. Should be public static int...Actually, it could be internal too. This is definitely not something we want to expose.

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (jimon) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs]: `// Now subsequent touches should not be ignored`
:)

File: Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs:271]: If we remove sorting by monitorIndex, which includes the binding index in the lowest bits, we'll end up processing bindings in potentially a different order than they are specified in the input asset. I can't decide if that's ok or not. Wasn't there some discussion recently in devs-input about a team using the explicit order of bindings to fall through to less specific bindings if more specific ones weren't available? Can this break that?

**Review by andrew-oc — CHANGES_REQUESTED** [review]

File: Assets/Tests/InputSystem/CoreTests_Actions.cs
Comment (jamesmcgill) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: Good catch thanks!

File: Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs
Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs:271]: The sorting was merged on the 24th March and it effectively was reversing the order they would have been in before. So bindings with the sorting enabled were now in descending order e.g. 4,3,2,1. This PR change would keep them in whatever order they were before 24th March (which I think would be ascending). 

I'm not sure about the order in the asset or about the binding fall through discussion, but I'm thinking unless these things were post 24th March, then they might have also been broken by adding the sort, and this PR should probably help restore those too?

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs
Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs]: Cheers for the review!
Yeah looks like I forgot to change them after copying them. Good catch.

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs
Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs]: Yep, thanks for pointing that out

File: Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs:271]: XR team is using it for this use-case and might have a project to test with, contact @chris-massie may provide insight

Comment (andrew-oc) [reply Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs:271]: Yep, good point.

**Review by andrew-oc — APPROVED** [review]

**Pull request merged**
